HTML API: Add tests for comments

sirreal opened the pull request:
Trac ticket: Core-61530

Add tests for comment handling.

These tests caught a regression during the 6.6 development cycle (with processing instruction comment tag name) that was fixed, but the tests were not merged.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Add tests for comment types

demonstrates regression from push/pop

sirreal committed:
Update comment tests with a data provider

sirreal commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,72 @@
+	/**
+	 * Ensures that different types of comments are processed correctly.
+	 *
+	 * @ticket TBD
```
```suggestion
	 * @ticket 61530
```

sirreal committed:
Update ticket

aaronjorbin commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,72 @@
+	 *
+	 * @dataProvider data_comments
+	 */
+	public function test_comment_processing( string $html, int $token_position, string $expected_token_name, ?string $expected_comment_type, string $expected_modifiable_text = '', string $expected_tag = null ) {
```
It feels a bit awkward to have the order of the final three be:

nullable, default, string who default is null

It also looks like there is only one case where `$expected_modifiable_text` is set but `$expected_comment_type` is not. For consistency, I wonder if it would make sense to swap those two and make $expected_comment_type default to null?

aaronjorbin commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,72 @@
+			'Normative comment'              => array( '<!-- A comment. -->', 1, '#comment', WP_HTML_Processor::COMMENT_AS_HTML_COMMENT, ' A comment. ' ),
+			'Abruptly closed comment'        => array( '<!-->', 1, '#comment', WP_HTML_Processor::COMMENT_AS_ABRUPTLY_CLOSED_COMMENT ),
+			'Invalid HTML comment !'         => array( '<! Bang opener >', 1, '#comment', WP_HTML_Processor::COMMENT_AS_INVALID_HTML, ' Bang opener ' ),
+			'Invalid HTML comment ?'         => array( '<? Question opener >', 1, '#comment', WP_HTML_Processor::COMMENT_AS_INVALID_HTML, ' Question opener ' ),
```
Is it worth adding a string with `<?PHP` for when someone tries to parse PHP or do you think that is caught by the above one and is unnecessary?

sirreal replied:
It's covered, but there's no harm in adding `<?php` as well.

sirreal committed:
Apply suggestions on data provider arg ordering

sirreal committed:
Add PHP processing instruction comment test

sirreal commented:
Good suggestions, thanks @aaronjorbin. I've applied them.

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,73 @@
+
+		for ( $i = 0; $i < $token_position; $i++ ) {
+			$processor->next_token();
+		}
```
given that all of these start in the first position, can we remove `$token_position`? often I try to be descriptive in the argument name, so instead of `$html` here, we could have `$html_starting_with_comment`

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,73 @@
+		$this->assertSame(
+			$expected_token_name,
+			$processor->get_token_name(),
+			"Found incorrect token name {$processor->get_token_name()}."
```
Keep in mind that the generated message will already say something like this:

```
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'#comment'
+'#funky-comment'
```

So I don't know if our message here is helping the clarity of the output. In some cases it simply seems needless to add the interpolated string, in others I might lean on saying something like `"Failed to identify token as a {$expected_token_name}"` rather than showing again the wrong value.

I want these messages, inasmuch as is possible, to be without noise, and lead people to know how to fix the code, or point out exactly what's wrong.

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessorComments.php:
```diff
@@ -0,0 +1,73 @@
+	 *
+	 * @dataProvider data_comments
+	 */
+	public function test_comment_processing( string $html, int $token_position, string $expected_token_name, string $expected_modifiable_text = '', string $expected_comment_type = null, string $expected_tag = null ) {
```
this is a really long function signature and I think it could be split into two functions, the `$token_position` eliminated, and all of the optional values removed.

Apart from funky comments everything under test is a comment, which is in the test name, so we could then eliminate the `$expected_token_name` and simply assert that the first token is a `#comment`.

For the funky comment test it would be the same but for `#funky-comment`.

sirreal committed:
Remove redundant token position and iteration

sirreal committed:
Remove redundant failure message

sirreal committed:
Split into simplified funky/regular comment tests

Use two different tests with much simpler function signatures.

sirreal commented:
Thanks @dmsnell, I applied those suggestions and cleaned things up. The signatures do become much simpler when we split the `#comment` and `#funky-comment` tests.

sirreal committed:
fixup! Split into simplified funky/regular comment tests

dmsnell reviewed (approved):
These are perfect. I don't like, in hindsight, that `get_tag()` returns a value for these. Too late to change that, perhaps. Either way, not an issue for this PR. Thanks for adding the tests.

dmsnell commented:
Merged in [[58734]](https://core.trac.wordpress.org/changeset/58734)
https://github.com/wordpress/wordpress-develop/commit/bd61b0bf236ed9c04bbbb50fd9b34d1b94970c3c
